templates/devshells/haskell/builders.ncl

```diff
@@ -0,0 +1,281 @@
+  },
+
+  # nodePackages_latest.typescript-language-server is marked broken in nixpkgs
+  JavascriptShell = BashShell & {
+    inputs_spec | {
+      nodejs.input | priority -100 = "nixpkgs",
+      nodePackages_latest.typescript-language-server.input | priority -100 = "nixpkgs",
+      ..
+    },
+    inputs,
+    output.structured_env = {
+      PATH = [
+        s%"%{inputs.bash}/bin"%,
+        s%"%{inputs.nodejs}/bin"%,
+        s%"%{inputs.nodePackages_latest.typescript-language-server}/bin"%,
+      ],
+    },
+    output.env = record.map (fun _n xs => array.foldl (fun acc x => s%"%{acc}:%{x}"%) "" xs) output.structured_env,
+  },
+
+  RacketShell = BashShell & {
+    inputs_spec | {
+      racket.input | priority -100 = "nixpkgs",
+      ..
+    },
+    inputs,
+    output.structured_env = {
+      PATH = [
+        s%"%{inputs.bash}/bin"%,
+        s%"%{inputs.racket}/bin"%,
+      ],
+    },
+    output.env = record.map (fun _n xs => array.foldl (fun acc x => s%"%{acc}:%{x}"%) "" xs) output.structured_env,
+  },
+
+  ScalaShell = BashShell & {
+    inputs_spec | {
+      scala.input | priority -100 = "nixpkgs",
+      metals.input | priority -100 = "nixpkgs",
+      ..
+    },
+    inputs,
+    output.structured_env = {
+      PATH = [
+        s%"%{inputs.bash}/bin"%,
+        s%"%{inputs.scala}/bin"%,
+        s%"%{inputs.metals}/bin"%,
+      ],
+    },
+    output.env = record.map (fun _n xs => array.foldl (fun acc x => s%"%{acc}:%{x}"%) "" xs) output.structured_env,
+  },
+
+  # broken: PyLTI-0.7.0 not supported for interpreter python3.10
+  # works with a regular nix shell though
+  Python310Shell = BashShell & {
+    inputs_spec | {
+      python310.input | priority -100 = "nixpkgs",
+      python310Packages.python-lsp-server.input | priority -100 = "nixpkgs",
+      ..
+    },
+    inputs,
+    output.structured_env = {
+      PATH = [
+        s%"%{inputs.bash}/bin"%,
+        s%"%{inputs.python310}/bin"%,
+        s%"%{inputs.python310Packages.python-lsp-server}/bin"%,
+      ],
+    },
+    output.env = record.map (fun _n xs => array.foldl (fun acc x => s%"%{acc}:%{x}"%) "" xs) output.structured_env,
+  },
+
+  ErlangShell = BashShell & {
+    inputs_spec | {
+      erlang.input | priority -100 = "nixpkgs",
+      erlang-ls.input | priority -100 = "nixpkgs",
+      ..
+    },
+    inputs,
+    output.structured_env = {
+      PATH = [
+        s%"%{inputs.bash}/bin"%,
+        s%"%{inputs.erlang}/bin"%,
+        s%"%{inputs.erlang-ls}/bin"%,
+      ],
+    },
+    output.env = record.map (fun _n xs => array.foldl (fun acc x => s%"%{acc}:%{x}"%) "" xs) output.structured_env,
+  },
+
+  HaskellShell = BashShell & {
+    inputs_spec | {
+      cabal-install.input | priority -100 = "nixpkgs",
+      ghc.input | priority -100 = "nixpkgs",
+      haskell-language-server.input | priority -100 = "nixpkgs",
+      stack.input | priority -100 = "nixpkgs",
```

Comment:
Why is the priority `-100` and not `default`?

Reply:
I _think_ when we first made the shells we had a stupid issue like the basic `NickelPackage` contract defines a default somewhere and then we have two conflicting default, so we had to make the priority "one layer up, but still easily overridable". I would say that this is orthogonal to the Haskell shell in particular, but in parallel we will definitely revisit this to see what is the issue again and if we can get rid of those pesky `-100`.

Reply:
@aspiwack the file `builders.ncl` is the library containing all the builders, then in your nickel file you can merge all the builders you want to create your shell.

Reply:
@rapenne-s is `builders.ncl` copied in every template then?

Reply:
@aspiwack Getting haskell right is hard that's true (... in general, with or without nix! :P). Happy to implement Julien's  wrapper. Regarding the GHC version, the pragmatic approach would be to allow the user to choose what version in a simple way. Then it's up to the user making sure the choosen LTS matches the GHC version (if they use stack).

Would this work for you?

Reply:
As soon as the ghc version is overridable, then it's easy to make sure that everything is compatible in the style of my Nix shell above (well, except the Stackage snapshot, but I think this is pretty unavoidable).

Reply:
@aspiwack yes, to be ready for FOSDEM we made things easier, we have the rust example containing a script creating all the other templates using the rust as a template itself.

We certainly want something better in the future, but it's still not clear how we want to distribute Nickel-nix to users (or its installation process).

Reply:
Understood.